transformer_engine/pytorch/ops/op.py

```diff
@@ -0,0 +1,428 @@
+
+            # Make sure FP8 metadata matches FP8 autocast context
+            # if any(fp8_meta is not None for fp8_meta in self._fp8_metas):
+            for fp8_meta in self._fp8_metas.values():
+                self._check_fp8_meta(fp8_meta)
+
+            # Register FP8 metadata for amax and scale update
+            if not FP8GlobalStateManager.fp8_graph_capturing():
+                if self.num_fp8_scales("input"):
+                    FP8GlobalStateManager.add_fp8_tensors_to_global_buffer(
+                        self.get_fp8_meta("input"),
+                    )
+                if self.num_fp8_scales("param"):
+                    fp8_params = list(filter(is_float8_tensor, self.parameters()))
+                    FP8GlobalStateManager.add_fp8_tensors_to_global_buffer(
+                        self.get_fp8_meta("param"),
+                        fp8_weights=(fp8_params if fp8_params else None),
+                    )
+                if self.num_fp8_scales("grad_output"):
+                    FP8GlobalStateManager.add_fp8_tensors_to_global_buffer(
+                        self.get_fp8_meta("grad_output"),
+                    )
+
+    @abc.abstractmethod
+    def op_forward(
+        self,
+        ctx: OperationContext,
+        input: torch.Tensor,  # pylint: disable=redefined-builtin
+        prev_op: Optional[BasicOperation] = None,
+        next_op: Optional[BasicOperation] = None,
+        **kwargs: Any,
+    ) -> torch.Tensor:
+        """Forward pass
+
+        Parameters
+        ----------
+        ctx: OperationContext
+            Context to coordinate between forward and backward passes
+        input: torch.Tensor
+            Input tensor
+
+        Returns
+        -------
+        torch.Tensor:
+            Output tensor
+
+        """
+
+    @abc.abstractmethod
+    def op_backward(
+        self,
+        ctx: OperationContext,
+        grad_output: torch.Tensor,
+    ) -> tuple[torch.Tensor, Iterable[Optional[torch.Tensor]]]:
+        """Backward pass
+
+        Parameters
+        ----------
+        ctx: OperationContext
+            Context to coordinate between forward and backward passes
+        grad_output: torch.Tensor
+            Loss gradient w.r.t. operation output
+
+        Returns
+        -------
+        torch.Tensor
+            Loss gradient w.r.t. operation input
+        Iterable of torch.Tensor:
+            Loss gradients w.r.t. parameters
+
+        """
+
+    def fuser_forward(
+        self,
+        basic_op_ctxs: list[OperationContext],
+        input_: torch.Tensor,
+        basic_op_prev_ops: list[Optional[BasicOperation]],
+        basic_op_next_ops: list[Optional[BasicOperation]],
+        basic_op_kwargs: list[dict[str, Any]],
+    ) -> torch.Tensor:
+        return self.op_forward(
+            basic_op_ctxs[0],
+            input_,
+            basic_op_prev_ops[0],
+            basic_op_next_ops[0],
+            **basic_op_kwargs[0],
```

Comment:
Could you explain why we index 0 here?

Reply:
`OperationFuser` doesn't make any distinction between `BasicOperation` or `FusedOperation`, but interacts with them via the base class (e.g. `FusableOperation.fuser_forward`). A `FusableOperation` consists of one or more `BasicOperation`s, so a `BasicOperation` will recieve just one ctx from `OperationFuser` while `FusedOperation` may recieve multiple.